Lib/test/test_shutil.py

```diff
@@ -1665,6 +1665,49 @@ def test_register_archive_format(self):
         formats = [name for name, params in get_archive_formats()]
         self.assertNotIn('xxx', formats)
 
+    def _unlink_existing_file(self, path):
+        try:
+            os.unlink(path)
+        except FileNotFoundError:
+            pass
+
+    def test_make_tarfile_rootdir_nodir(self):
+        # GH-99203
+        self.addCleanup(self._unlink_existing_file, f'{TESTFN}.tar')
+        for dry_run in (0, True):
+            with self.subTest(dry_run=dry_run):
+                tmp_fd, tmp_file = tempfile.mkstemp(dir=self.mkdtemp())
```

Comment:
It leaks a file and a directory if the test fails.

Also, no need to generate an unique name for temporary file in a freshly created temporary directory. You can just use `os.path.join(self.mkdtemp(), 'file')` and `os.path.join(self.mkdtemp(), 'nonexisting')`.

Reply:
> It leaks a file and a directory if the test fails.

`BaseTest.mkdtemp` does `self.addCleanup(os_helper.rmtree, d)`, so the dir (with the file in it) shouldn't leak.

> Also, no need to generate an unique name for temporary file in a freshly created temporary directory. You can just use `os.path.join(self.mkdtemp(), 'file')` and `os.path.join(self.mkdtemp(), 'nonexisting')`.

I kept using `tempfile.mkstemp` for the `NotADirectoryError` test: if I want to set the filename, it looks like I'd need to call `open()` to create a file and use a context manager or call `close()` (`os.mknod` does not support some platforms). I think the filename being unique shouldn't hurt here. I see no advantage in changing this case as well?